### PR TITLE
zeropool-2-snjax

### PR DESCRIPTION
**Pull request by snjax**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/208

### CODE REVIEW

**Comment by Noc2** [comment]
Thanks for the delivery. We will look into it as soon as possible.

**Comment by semuelle** [comment]
Hey @snjax, milestone deliveries should use the [delivery template](https://github.com/w3f/Grant-Milestone-Delivery#mailbox-milestone-delivery-process) (you used our evaluation template) and go into the `deliveries` folder. Could you fix that, please?

File: deliveries/zeropool-2-snjax.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Evaluation
+
+
+
+- **Status:** Accepted
```
Comment (mmagician) [inline deliveries/zeropool-2-snjax.md:5]:
```suggestion
```

Comment (mmagician) [reply deliveries/zeropool-2-snjax.md:5]: That's reserved for evaluations ;)

**Comment by mmagician** [comment]
Let me know once you've got the tutorial ready, so I know where to start! I guess a lot has changed since Milestone 1.

File: deliveries/zeropool-2-snjax.md (hunk 1 above)
Comment (semuelle) [reply deliveries/zeropool-2-snjax.md:5]: @snjax, I submitted a PR to your fork to adapt the application to the right template. This issue will be resolved if you merge it.

Comment (snjax) [reply deliveries/zeropool-2-snjax.md]: Thank you for the PR.

**Comment by mmagician** [comment]
@snjax I had a quick look at your delivery. I'm still waiting for the complete tutorial.
In addition, there is no deliverable showing how your libraries integrate with substrate. According to your initial grant application:
> We will create a library for proving and verifying private transactions, compatible with the substrate pallet

The compatibility is not yet demonstrated in this milestone.

**Comment by snjax** [comment]
@mmagician, our current implementation is compatible with substrate, we use no_std rust for parts of fawkes-crypto, used to implement the substrate module (like final field math).

I could show you the substrate module as part of milestone 3 this month.

Also, currently, we have a live EVM version on [kovan](https://kovan.testnet.console.v2.zeropool.network/) with custom, not EVM-specific [ABI](https://github.com/zeropoolnetwork/pool-evm-single-l1/blob/main/contracts/Parameters.sol#L25).

We are working on a substrate module, supporting this ABI.

**Comment by mmagician** [comment]
@snjax I see your point that since your library is `no_std` compatible, the integration with substrate would be straightforward. However, since that integration is not explicitly demonstrated in M2, it might be easier to wait for M3 as you suggest and assess both milestones in parallel. What do you think?

**Comment by snjax** [comment]
@mmagician, assessing both milestones in parallel would be much more convenient for us.

**Comment by snjax** [comment]
@alxs the work is in progress. We were focused on more common non-substrate-specific points.

[Here](https://github.com/zeropoolnetwork/zeropool-substrate/tree/main/pallets/zeropool-pallet) is current WIP repository for substrate. It seems this month we are able to show you the final result (like our evm kovan testnet).

**Comment by semuelle** [comment]
> It seems this month we are able to show you the final result (like our evm kovan testnet).

Could you provide an update, @snjax? Does it make sense to keep this PR open if you want to finish and submit M3 alongside?

**Comment by snjax** [comment]
@semuelle We have built it. [Here](https://onedrive.live.com/?authkey=%21AJYZU6QsXBIWF9w&id=FFBBD74F1BEF0EE8%212567) is the screencast of deposit, transfer, and withdrawal implemented on substrate tech stack and ZeroPool circuit. 

Now we are working on documentation and deployment.

I will push an update into this PR soon.

**Comment by semuelle** [comment]
Hey @snjax, we are still waiting for an update from your side. How is it looking, any news since your last comment?

**Comment by snjax** [comment]
Hi, @semuelle . We have deployed the testnet at https://substrate.testnet.console.v2.zeropool.network/ . Currently, we are finalizing the documentation of the project.

**Comment by snjax** [comment]
@semuelle We have updated the milestone delivery. So, please, have a look and we are waiting for your feedback.

**Comment by mmagician** [comment]
@snjax Apologies for taking so long. I've created a first version of the evaluation [in this PR](https://github.com/w3f/Grant-Milestone-Delivery/pull/484), where you'll find the comments I have so far.

**Comment by semuelle** [comment]
ping @snjax

**Comment by snjax** [comment]
@semuelle We received your comments and working on documentation now. You will get the update in August.

**Comment by alxs** [comment]
@snjax friendly reminder - when do you think you'll be able to get back to us?

**Comment by snjax** [comment]
Hi, @alxs.

The documentation is published [here](https://docs.zeropool.network/docs/implementation/account-and-notes/accounts) and [here](https://github.com/zeropoolnetwork/docs).

We will publish substrate-specific things soon, but cryptography, documented there, corresponds to our substrate implementation.

Currently, we are working on plonk proving backend, recursive prover (zkzkrollup), and non-evm implementation for our frontend (https://testnet.app.zeropool.network/).

**Comment by Noc2** [comment]
@snjax I just wanted to check in regarding the current status. Would you still like to finish the delivery?

**Review by Noc2 — CHANGES_REQUESTED** [review]
(see above)

**Comment by snjax** [comment]
@Noc2 More substrate-specific documentation is added [here](https://docs.zeropool.network/docs/implementation/contracts-and-circuits/substrate/).

We are currently working on switching to Plonk. But Groth16 implementation is stable.

Sure, I would like to finish the delivery.

**Comment by snjax** [comment]
@Noc2 @alxs Have you viewed the documentation? Have you got any comments?

**Comment by semuelle** [comment]
> Have you viewed the documentation? Have you got any comments?

Thanks for the ping. alxs has left and @Noc2 is currently ooo, but someone from our team is going to review it soon.

**Comment by snjax** [comment]
@semuelle 

the current video of using the frontend of our pallet is available here https://www.youtube.com/watch?v=YRJadaIrBJs

The web interface is deployed at https://substrate.testnet.console.v2.zeropool.network/

**Review by semuelle — COMMENTED** [review]
Hi @snjax. Sorry for the long pause. I have now started reviewing your delivery. 
When I run the ZeroPool node through docker (`docker-compose -f docker-compose.full.yml up`), the relayer stops with the following error message. Do you know what might be the reason?

```sh
relayer_polkadot    | thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/params.rs:30:36
relayer_polkadot    | note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
relayer_polkadot    | /app/zp-relayer/pool.js:32
relayer_polkadot    |         this.treeParams = libzeropool_rs_node_1.Params.fromFile('./params/tree_params.bin');
relayer_polkadot    |                                                        ^
relayer_polkadot    | 
relayer_polkadot    | Error: internal error in Neon module: called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
relayer_polkadot    |     at new Pool (/app/zp-relayer/pool.js:32:56)
relayer_polkadot    |     at Object.<anonymous> (/app/zp-relayer/pool.js:196:16)
relayer_polkadot    |     at Module._compile (node:internal/modules/cjs/loader:1103:14)
relayer_polkadot    |     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
relayer_polkadot    |     at Module.load (node:internal/modules/cjs/loader:981:32)
relayer_polkadot    |     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
relayer_polkadot    |     at Module.require (node:internal/modules/cjs/loader:1005:19)
relayer_polkadot    |     at require (node:internal/modules/cjs/helpers:102:18)
relayer_polkadot    |     at Object.<anonymous> (/app/zp-relayer/utils/redisFields.js:15:16)
relayer_polkadot    |     at Module._compile (node:internal/modules/cjs/loader:1103:14)
relayer_polkadot    | error Command failed with exit code 1.
relayer_polkadot    | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
relayer_polkadot exited with code 1
``

**Comment by snjax** [comment]
@semuelle 

This is because you run it with no parameters like proving and verifying keys. 

We added the description of parameters into the README of [zeropoolnetwork/zeropool-substrate](https://github.com/zeropoolnetwork/zeropool-substrate).

Also, you can get it from [here](https://github.com/zeropoolnetwork/zeropool-test-params).

**Review by semuelle — CHANGES_REQUESTED** [review]
> We added the description of parameters into the README of [zeropoolnetwork/zeropool-substrate](https://github.com/zeropoolnetwork/zeropool-substrate).

The path to the `params` folder in the docker-compose file wasn't quite right, but I got it to work. 

Unless I'm overlooking something, the article is still missing for both milestones. Could you still do that and update the delivery doc? Since you are submitting both milestones at once, one would be enough.

**Comment by semuelle** [comment]
ping @snjax

**Comment by snjax** [comment]
@semuelle We have published the overview&tutorial how to use ZeroPool privacy engine on substrate tech stack.

https://medium.com/zeropool/zeropool-implementing-privacy-preserving-transactions-with-zksnarks-and-substrate-1bd060505166

**Comment by snjax** [comment]
@semuelle We have published the overview&tutorial how to use ZeroPool privacy engine on substrate tech stack.

https://medium.com/zeropool/zeropool-implementing-privacy-preserving-transactions-with-zksnarks-and-substrate-1bd060505166

**Comment by semuelle** [comment]
Thanks, @snjax. I am happy to report that both or your milestones are hereby finally accepted. See evaluation notes [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/zeropool_1_mmagician-semuelle.md) and [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/zeropool_2_semuelle.md).

You will probably have to resend your invoices, as they require a somewhat recent date on them. :)

**Comment by snjax** [comment]
@semuelle Thanks for accepting all milestones. 
It's not very clear to me, what you mean. 
If you write about the accessibility of 0xb6F9F891497C0B72a8A817f3D3E3C721fca6f9CC in the submission there are no changes from my side and I have not sent any other invoices.

**Comment by semuelle** [comment]
@snjax, someone submitted an invoice through the [invoice form](https://forms.gle/LSRr7PCjBpEbKGh89) the same day this PR was submitted.

- The invoice is only for M2, we don't have one for M1.
- The invoice is from Dec 2021. Could you send one with a more recent date? 
- Also, the payment address doesn't match the application. Both addresses must match, so you either have to change it in the invoice (easiest) or submit an amendment to the contract.

**Comment by snjax** [comment]
@semuelle the invoice for Milestones 1 and 2 is sent.

**Comment by snjax** [comment]
@fededubbi, I have updated the invoice with the information you requested.

**Comment by RouvenP** [comment]
hi @snjax we transferred the payment today

**Comment by snjax** [comment]
@semuelle Thank you for the experience of cooperation with Web3 Foundation team. It was amazing.  Btw, are you at EthCC? It would be nice to meet up.

**Comment by Noc2** [comment]
Hi @snjax, @semuelle is currently OOO and not at EthCC, but @bhargavbh from the Web3 Foundation might be there. Other than that, feel free to reach out via email: david@web3.foundation and we can potentially schedule a call.

**Comment by snjax** [comment]
@Noc2 thank you for your reply. I will send email.

**Pull request merged**